### PR TITLE
[Parser] Simplify sequence expression parsing 

### PR DESCRIPTION
**Pull request by rintaro**
(Depends on https://github.com/apple/swift/pull/60731)

`SequenceExprSyntax` now gurantees that it has odd number elements using `UnresolvedTernaryExpr`/`UnresolvedAsExpr`/`UnresolvedIsExpr`.

Factored out the operator parsing into `parseSequenceExpressionOperator()` which returns the operator expresssion, and optionally the RHS, so we can avoid nested loops.
Also, `parseSequenceExprssion()` now has a performance optimization that avoids heap `Array` construction unless the expression is actually sequenced.

### CODE REVIEW

File: Sources/SwiftParser/Expressions.swift [hunk 1]
```diff
@@ -42,23 +42,6 @@ extension Parser {
 }
 
 extension Parser {
-  struct PendingTernary {
```
Comment (DougGregor) [inline Sources/SwiftParser/Expressions.swift:45]: I'm so glad to see this struct go away. The parser is much simpler now!

File: Sources/SwiftParser/Expressions.swift [hunk 2]
```diff
@@ -69,187 +52,190 @@ extension Parser {
...
+  ///
+  /// Grammar
+  /// =======
+  ///
+  ///     assignment-operator → '='
+  ///     conditional-operator → '?' expression ':'
+  ///     type-casting-operator → 'is' type
+  ///     type-casting-operator → 'as' type
+  ///     type-casting-operator → 'as' '?' type
+  ///     type-casting-operator → 'as' '!' type
```
Comment (DougGregor) [inline Sources/SwiftParser/Expressions.swift:138]: Do we need to document the arrow operator here?

**Review by DougGregor — APPROVED** [review]
This is fantastic, and a huge help for the operator-precedence library in https://github.com/apple/swift-syntax/pull/619.

**Comment by rintaro** [comment]
https://github.com/apple/swift/pull/60731
@swift-ci Please test

**Pull request merged**